--- Day 12: Digital Plumber ---

Walking along the memory banks of the stream, you find a small village that is experiencing a little confusion: some programs can't communicate with each other.

Programs in this village communicate using a fixed system of pipes. Messages are passed between programs using these pipes, but most programs aren't connected to each other directly. Instead, programs pass messages between each other until the message reaches the intended recipient.

For some reason, though, some of these messages aren't ever reaching their intended recipient, and the programs suspect that some pipes are missing. They would like you to investigate.

You walk through the village and record the ID of each program and the IDs with which it can communicate directly (your puzzle input). Each program has one or more programs with which it can communicate, and these pipes are bidirectional; if 8 says it can communicate with 11, then 11 will say it can communicate with 8.

You need to figure out how many programs are in the group that contains program ID 0.

For example, suppose you go door-to-door like a travelling salesman and record the following list:

```
0 <-> 2
1 <-> 1
2 <-> 0, 3, 4
3 <-> 2, 4
4 <-> 2, 3, 6
5 <-> 6
6 <-> 4, 5
```

In this example, the following programs are in the group that contains program ID 0:

Program 0 by definition.
Program 2, directly connected to program 0.
Program 3 via program 2.
Program 4 via program 2.
Program 5 via programs 6, then 4, then 2.
Program 6 via programs 4, then 2.
Therefore, a total of 6 programs are in this group; all but program 1, which has a pipe that connects it to itself.

How many programs are in the group that contains program ID 0?

In [2]:
test_data = [
    "0 <-> 2",
    "1 <-> 1",
    "2 <-> 0, 3, 4",
    "3 <-> 2, 4",
    "4 <-> 2, 3, 6",
    "5 <-> 6",
    "6 <-> 4, 5",
]

In [3]:
test_data

['0 <-> 2',
 '1 <-> 1',
 '2 <-> 0, 3, 4',
 '3 <-> 2, 4',
 '4 <-> 2, 3, 6',
 '5 <-> 6',
 '6 <-> 4, 5']

In [4]:
test_indices = [x.split(" ")[0] for x in test_data]
test_indices

['0', '1', '2', '3', '4', '5', '6']

In [5]:
test_values = [x.split(" <-> ")[1].split(", ") for x in test_data]
test_values

[['2'], ['1'], ['0', '3', '4'], ['2', '4'], ['2', '3', '6'], ['6'], ['4', '5']]

In [6]:
test_groups = {}
for i, j in enumerate(test_indices):
    test_groups[j] = test_values[i]
test_groups

{'0': ['2'],
 '1': ['1'],
 '2': ['0', '3', '4'],
 '3': ['2', '4'],
 '4': ['2', '3', '6'],
 '5': ['6'],
 '6': ['4', '5']}

In [7]:
list(test_groups.values())

[['2'], ['1'], ['0', '3', '4'], ['2', '4'], ['2', '3', '6'], ['6'], ['4', '5']]

In [8]:
def find_groups(index, groups):
    matches = [index]
    for i, group in enumerate(groups):
        if index in list(groups.values())[i]:
            matches.append(list(groups.keys())[i])
    return matches

In [9]:
test_in_group_zero = ['0']
for group in test_in_group_zero:
    matches = find_groups(group, test_groups)
    for m in matches:
        if m not in test_in_group_zero:
            test_in_group_zero.append(m)
print(set(test_in_group_zero))
len(set(test_in_group_zero))

{'3', '5', '6', '2', '4', '0'}


6

In [10]:
with open("day12.txt") as f:
    day12 = f.readlines()
    day12 = [x.strip("\n") for x in day12]

In [11]:
day12[0:3]

['0 <-> 199, 1774', '1 <-> 350, 1328, 1920', '2 <-> 477, 984, 1419']

In [12]:
indices = [x.split(" ")[0] for x in day12]
indices[0:3]

['0', '1', '2']

In [13]:
values = [x.split(" <-> ")[1].split(", ") for x in day12]
values[0:3]

[['199', '1774'], ['350', '1328', '1920'], ['477', '984', '1419']]

In [14]:
groups = {}
for i, j in enumerate(indices):
    groups[j] = values[i]
groups['0']

['199', '1774']

In [15]:
in_group_zero = ['0']
for group in in_group_zero:
    matches = find_groups(group, groups)
    for m in matches:
        if m not in in_group_zero:
            in_group_zero.append(m)
print(set(in_group_zero))
len(set(in_group_zero))

{'1077', '1954', '199', '1264', '1309', '53', '1554', '683', '1188', '1520', '1356', '941', '1854', '27', '1673', '956', '420', '1910', '1026', '588', '1870', '606', '47', '699', '1120', '461', '1197', '533', '64', '1846', '94', '438', '165', '385', '1559', '1933', '1045', '951', '504', '434', '215', '787', '1212', '11', '1552', '628', '1880', '134', '1412', '855', '1646', '1723', '1627', '85', '1222', '918', '287', '207', '1315', '745', '1597', '304', '1029', '722', '925', '296', '1882', '354', '1542', '1002', '1168', '1503', '940', '558', '1967', '338', '1165', '696', '1270', '1545', '625', '1415', '234', '1834', '1474', '1504', '830', '176', '1274', '1000', '1989', '68', '120', '391', '241', '1671', '1945', '1887', '1483', '608', '1450', '1342', '1087', '1004', '1875', '1362', '277', '1193', '1135', '1495', '152', '1792', '387', '1888', '301', '684', '839', '634', '1771', '1698', '819', '1470', '967', '56', '794', '920', '490', '1438', '412', '1263', '678', '540', '1152', '888', '17

175

That's the right answer! You are one gold star closer to debugging the printer.

--- Part Two ---

There are more programs than just the ones in the group containing program ID 0. The rest of them have no way of reaching that group, and still might have no way of reaching each other.

A group is a collection of programs that can all communicate via pipes either directly or indirectly. The programs you identified just a moment ago are all part of the same group. Now, they would like you to determine the total number of groups.

In the example above, there were 2 groups: one consisting of programs 0,2,3,4,5,6, and the other consisting solely of program 1.

How many groups are there in total?

In [120]:
def find_group(seed):
    group = {seed}
    new = {seed}
    while new:
        next_new = set()
        for item in new:
            next_new.update(groups[item])
        new = next_new - group
        group.update(next_new)
    return group

In [124]:
counter = 0
remaining = set(groups)
while remaining:
    counter += 1
    group = find_group(remaining.pop())
    remaining -= group
counter

213

Your puzzle answer was 213.

That's the right answer! You are one gold star closer to debugging the printer.